### PR TITLE
proposal: Expose single annotations via downward API

### PR DESCRIPTION
**Pull request by nhlfr**
Ref #31218

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36058)

### CODE REVIEW

**Comment by nhlfr** [comment]
@smarterclayton @pmorie

File: docs/proposals/annotations-downward-api.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+annotation seems to be a valid way to do this. However, to finally pass this
+information to the containerized application, we need to pass this annotation.
+That's why the downward API for annotations is needed here.
+
+## API
+
+The exact `fieldPath` to the annotation will look like:
+
+```
+metadata.annotations['annotationKey']
```
Comment (thockin) [inline docs/proposals/annotations-downward-api.md]: Why the choice of single-quotes?  That is neither JSON nor Go, and seems to have weird meaning in YAML.  Do we need quotes here at all?  If so, we should use double-quotes.

File: docs/proposals/annotations-downward-api.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        fieldPath: metadata.annotations['pod.beta.kubernetes.io/petset-index']
+```
+
+## Implementation
+
+In general, this environment downward API part will be implemented in the same
+place as the other metadata - as a label conversion function.
+
+The most tricky part of implementation will be proper validation of the `fieldPath`.
+We shouldn't allow any of these characters `\`, `'`, `[`, `]` to be a part of
```
Comment (thockin) [inline docs/proposals/annotations-downward-api.md]: this is already true, unless I am missing somethign?

File: docs/proposals/annotations-downward-api.md (hunk 1 above)
Comment (smarterclayton) [inline docs/proposals/annotations-downward-api.md]: Not just talking about annotations.  Any key applies.  I don't think JSONPath demands single-quotes, but all examples I see are canonically single quote.

**Comment by thockin** [comment]
Well, sure, you can put any string you like there, but it is not going to
work.

I don't know where to find docs on JSONPath and I don't know it as well as
go-template, but here's what I see:

```
$ ./_output/bin/kubectl get nodes -o
jsonpath='{.items[*].metadata.annotations}'
map[volumes.kubernetes.io/controller-managed-attach-detach:true] map[
volumes.kubernetes.io/controller-managed-attach-detach:true] map[
volumes.kubernetes.io/controller-managed-attach-detach:true] map[
volumes.kubernetes.io/controller-managed-attach-detach:true]
```

**Comment by nhlfr** [comment]
I'm going to move it to kubernetes/community.

**Pull request closed without merging**